skimage/_shared/tests/test_utils.py

```diff
@@ -347,3 +349,318 @@ def test_deprecate_func():
+            # if PYTHONOPTIMIZE is set to 2, docstrings are stripped
+            assert (
+                _func_deprecated_params.__doc__
+                == """Expected docstring.
+
+
+    Parameters
+    ----------
+    arg0 : int
+        First unchanged parameter.
+    arg1 : int, optional
+        Second unchanged parameter.
+
+    Other Parameters
+    ----------------
+    old0 : DEPRECATED
+        `old0` is deprecated.
+
+        .. deprecated:: 0.10
+    old1 : DEPRECATED
+        `old1` is deprecated.
+
+        .. deprecated:: 0.10
+"""
+            )
+
+    def test_replace_docstring(self):
+        assert _func_replace_params.__name__ == "_func_replace_params"
+        if sys.flags.optimize < 2:
+            # if PYTHONOPTIMIZE is set to 2, docstrings are stripped
+            assert (
+                _func_replace_params.__doc__
+                == """Expected docstring.
+
+
+    Parameters
+    ----------
+    arg0 : int
+        First unchanged parameter.
+    new0 : int, optional
+        First new parameter.
+
+        .. versionadded:: 0.10
+    new1 : int, optional
+        Second new parameter.
+
+        .. versionadded:: 0.10
+    arg1 : int, optional
+        Second unchanged parameter.
+
+    Other Parameters
+    ----------------
+    old0 : DEPRECATED
+        Deprecated in favor of `new1`.
+
+        .. deprecated:: 0.10
+    old1 : DEPRECATED
+        Deprecated in favor of `new0`.
+
+        .. deprecated:: 0.10
+"""
+            )
+
+    def test_remove_warning(self):
+        match = (
+            r".*`old[01]` is deprecated since version 0\.10 and will be removed "
+            r"in 0\.12.* see the documentation of .*_func_deprecated_params`."
+        )
+        with pytest.warns(FutureWarning, match=match):
+            assert _func_deprecated_params(1, 2) == (1, DEPRECATED, DEPRECATED, None)
+        with pytest.warns(FutureWarning, match=match):
+            assert _func_deprecated_params(1, 2, 3) == (1, DEPRECATED, DEPRECATED, None)
+        with pytest.warns(FutureWarning, match=match):
+            assert _func_deprecated_params(1, old0=2) == (
+                1,
+                DEPRECATED,
+                DEPRECATED,
+                None,
+            )
+        with pytest.warns(FutureWarning, match=match):
+            assert _func_deprecated_params(1, old1=2) == (
+                1,
+                DEPRECATED,
+                DEPRECATED,
+                None,
+            )
+
+        with warnings.catch_warnings(record=True) as record:
+            assert _func_deprecated_params(1, arg1=3) == (1, DEPRECATED, DEPRECATED, 3)
+        assert len(record) == 0
+
+    def test_replace_warning(self):
+        match = (
+            r".*`old[0,1]` is deprecated since version 0\.10 and will be removed "
+            r"in 0\.12.* see the documentation of .*_func_replace_params`."
+        )
+
+        with pytest.warns(FutureWarning, match=match):
+            assert _func_replace_params(1, 2) == (
```

Comment:
Ha, positional arguments can definitely be evil! :sweat_smile:

Is `new1` getting the `2` value because `new1` is replacing `old0` which, position-wise, is the parameter coming right after `arg0`?

Reply:
Yes. Among other things, I am also testing here that `deprecate_parameter` moves the value from `old0` to `new1` and replaces the former with `DEPRECATED`. This ensures that we don't forget some reference to `old0` somewhere that we will only notice when we complete the deprecation. It should also make completing the deprecation easier. The code responsible for this is

https://github.com/scikit-image/scikit-image/blob/5556066484c597cbbad346cecd8444838b0ab2b0/skimage/_shared/utils.py#L286-L291

and

https://github.com/scikit-image/scikit-image/blob/5556066484c597cbbad346cecd8444838b0ab2b0/skimage/_shared/utils.py#L317-L319